Update metadata editor to use JupyterLab's new form editor

marthacryan opened the pull request:
This PR adds the package [rjsf](https://react-jsonschema-form.readthedocs.io/)(react JSON schema form - a renderer for showing JSON schema as a form) to the metadata editor for consistency with JupyterLab (moved to using this package in the most recent release with jupyterlab/jupyterlab#11079. Going forward, this will also handle many of the more complex cases for rendering the metadata editor and provide built-in features that will solves issues such as #2269 and #2268.

Screenshots / gifs to come

**Todo:**
- [x] Still need to address validation failing on some categories and preventing save and close
~~- [ ] Update documentation~~ Follow up PR to come
- [x] Clean up look and feel with CSS

Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

karlaspuldaro commented on packages/metadata-common/src/MetadataEditor.tsx:
```diff
@@ -31,306 +24,434 @@ import {
-            {`"${this.displayName}" has unsaved changes, close without saving?`}{' '}
-          </p>
-        ),
+        body: <p>Metadata has unsaved changes, close without saving?</p>,
```
Maybe we should keep the reference to the metadata display name here

karlaspuldaro reviewed (commented):
This looks great! Only have a few comments/questions.

- Not introduced by this PR, but we should probably take a look at the warnings from Code QL when using `setState` in `MetadataEditorTags.tsx`:
```
React component state updates using setState may asynchronously update this.props and this.state, thus it is not safe to use either of the two when calculating the new state passed to setState.

Recommendation
Use the callback-based variant of setState: instead of calculating the new state directly and passing it to setState, pass a callback function that calculates the new state when the update is about to be performed.
```
- `MetadataEditor.tsx` file is pretty long, I wonder if it would make sense to move the widget code into its own file?

- In schemas updates I'm not sure why sometimes we completely remove `field_type: dropdown` even when it is an enum or a string array of more than one item, wouldn't those need to be updated to `ui:field: dropdown` ?

karlaspuldaro commented:
A few comments on the UI, design can be further discussed with the team
![image](https://user-images.githubusercontent.com/25207344/167218708-9c499838-4b9b-4bcf-9c12-751d1cf11581.png)

marthacryan commented:
Thanks for the review @karlaspuldaro!

> * Not introduced by this PR, but we should probably take a look at the warnings from Code QL when using `setState` in `MetadataEditorTags.tsx`:
> 
> ```
> React component state updates using setState may asynchronously update this.props and this.state, thus it is not safe to use either of the two when calculating the new state passed to setState.
> 
> Recommendation
> Use the callback-based variant of setState: instead of calculating the new state directly and passing it to setState, pass a callback function that calculates the new state when the update is about to be performed.
> ```

I'll try to fix this in this PR but we could also make an issue if it ends up being complicated. 

> * `MetadataEditor.tsx` file is pretty long, I wonder if it would make sense to move the widget code into its own file?

Good idea! It is a huge file, I'll move the widget into its own file. 

> * In schemas updates I'm not sure why sometimes we completely remove `field_type: dropdown` even when it is an enum or a string array of more than one item, wouldn't those need to be updated to `ui:field: dropdown` ?

The form editor automatically detects if there is an `enum` field defined in the schema and uses the dropdown component if so - so we no longer needed that UI hint because it's already figuring it out for us.

ajbozarth commented on packages/ui-components/src/FormComponents/CodeBlock.tsx:
```diff
@@ -0,0 +1,78 @@
+    return (): void => {
+      editorRef.current?.model.value.changed.disconnect(handleChange);
+    };
+    // NOTE: The parent component is unstable so props change frequently causing
+    // new editors to be created unnecessarily. This effect on mount should only
+    // run on mount. Keep in mind this could have side effects, for example if
+    // the `onChange` callback actually does change.
```
out of curiosity would it be worth adding a partial list of dependencies here for fields that we don't expect to change but we'd want to trigger an update? (ie fields not touched by the unstable parent component)

ajbozarth commented on packages/ui-components/src/FormComponents/MetadataEditorTags.tsx:
```diff
@@ -197,63 +198,79 @@ export class MetadataEditorTags extends React.Component<
         </button>
       );
     return (
```
We should address the codeQL warnings above, I think they can be addressed by some refactoring to use `useEffect` or similar

ajbozarth commented on packages/ui-components/src/icons.tsx:
```diff
@@ -99,6 +101,14 @@ export const pyIcon = new LabIcon({
+export const viewOffIcon = new LabIcon({
+  name: 'elyra:viewOff',
+  svgstr: viewOffIconSvg
+});
```
what lib did we get these icon from? do we need to add a license reference for them?

marthacryan replied:
It comes from carbon components (I think that's where we've gotten our other icons so it should be fine?)

ajbozarth commented on packages/metadata-common/package.json:
```diff
@@ -43,6 +43,7 @@
     "@lumino/messaging": "^1.4.3",
     "@lumino/widgets": "^1.19.0",
     "@material-ui/core": "^4.12.1",
+    "@rjsf/core": "^3.1.0",
```
this is only added here but should also be in ui-components since it is referenced by the components as well. Though if we move the FormEditor to ui-components it would only need to be referenced there

ajbozarth reviewed (commented):
Finally found time for a full code review. Mostly just nits and questions, nothing truly blocking so no requests for changes or approval for now.

I'll be testing it locally on another day, though on my initial start I seem to not be able to open the editor, with jupyterlab hanging with the Loading text and no errors in the console or log

marthacryan commented on elyra/metadata/schemas/kfp.json:
```diff
@@ -153,7 +150,8 @@
           "uihints": {
-            "secure": true,
+            "ui:field": "password",
+            "minLength": 8,
```
Oh weird

marthacryan commented on packages/metadata-common/src/FormEditor.tsx:
```diff
@@ -0,0 +1,189 @@
+ * Creates a uiSchema from given uihints and passes relevant information
+ * to the custom renderers.
+ */
+export const FormEditor: React.FC<IFormEditorProps> = ({
```
Good point, then I can move all rjsf code to ui-components. Moved it but I have to move the css over now

marthacryan commented on packages/metadata-common/src/MetadataEditorWidget.tsx:
```diff
@@ -0,0 +1,337 @@
+            }
+          },
+          required: ['display_name']
+        }
```
I actually added translation for this title / description since it's hard coded - does that work?

marthacryan commented on packages/metadata/src/index.ts:
```diff
@@ -44,17 +55,30 @@ const commandIDs = {
     editorServices: IEditorServices,
     status: ILabStatus,
+    componentRegistry: IFormComponentRegistry,
+    translator: ITranslator,
```
Added some more - not sure if it's covering every possible case but it's mostly there for the editor part

akchinSTC commented:
@marthacryan - played around with the combining the regex patterns in the kfp.json schema. 
Let me know if this works.
```
"pattern": "^[^ ]+( +[^ ]+)*$"
```

ajbozarth commented on packages/ui-components/package.json:
```diff
@@ -40,7 +40,8 @@
     "react-dom": "^17.0.1",
-    "react-json-tree": "^0.11.2"
+    "react-json-tree": "^0.11.2",
+    "@rjsf/core": "^3.1.0"
```
I believe this should be above react (the @ is sorted first)

ajbozarth commented on packages/ui-components/src/FormComponents/MetadataEditorTags.tsx:
```diff
@@ -137,130 +106,121 @@ export class MetadataEditorTags extends React.Component<
+                    <button
+                      onClick={handleClick}
+                      className={`${METADATA_EDITOR_TAG} tag unapplied-tag`}
+                      id={'editor' + '-' + tag + '-' + index}
```
why not use formatted strings for the strings in the return block (instead of using +)

ajbozarth commented on packages/ui-components/style/formeditor.css:
```diff
@@ -0,0 +1,284 @@
+  padding-bottom: 0;
+}
+
+.elyra-metadataEditor
```
we shouldn't have css specific to metadataEditior in ui-components, so either this should move back, or if it's actually generic to the form editor we should update the css class name (this applies to all the references in this file)

ajbozarth reviewed (changes requested):
a few comments on the code changes since my last review, and I'm building this to test locally atm

ajbozarth commented:
UI/UX Review as mentioned above, all tested in Safari:

- The mouseover description shows when mousing over both the title and input area for a filed as well as if you mouse over the description itself. This hides the title for the field below. Unless there's a reason to read it while filling out the field, or to copy/paste the descriptions contents I think a better UX would be to only show when mousing over the title. (@ptitzler option on this?) (it can be seen in the video in bullet 3)
- Show hide icon css not quite right
![Screen Shot 2022-05-23 at 4 09 49 PM](https://user-images.githubusercontent.com/13952758/169906551-059b3ad8-6298-44a4-a942-9eb7637a056d.png)
- Form spacing changes on initial mouseover, see video below:

https://user-images.githubusercontent.com/13952758/169907212-566a1a35-5647-4cdb-b899-ebe4c24042d8.mov

- some escaping is going wrong here:
![Screen Shot 2022-05-23 at 4 19 03 PM](https://user-images.githubusercontent.com/13952758/169907730-bf282e89-ff88-4ae3-914d-159602de5d69.png)
- description tooltips are misaligned for code snippet fields, also there is no description for tags (dont know if that's intentional) 
![Screen Shot 2022-05-23 at 4 20 43 PM](https://user-images.githubusercontent.com/13952758/169907930-07c5f761-f811-41b3-a980-d3f5fca736da.png)
![Screen Shot 2022-05-23 at 4 20 47 PM](https://user-images.githubusercontent.com/13952758/169907933-f35e0587-b3d2-401a-8802-0260b65d7f3f.png)
- css for this field looks off 
![Screen Shot 2022-05-23 at 4 22 13 PM](https://user-images.githubusercontent.com/13952758/169908121-2fc0b111-a924-4104-b124-bea57e0949d4.png)
- there's no longer button highlighting when mousing over the save button like before 

That's all I found on an initial pass, I only played with it for less than an hour

marthacryan commented:
> * The mouseover description shows when mousing over both the title and input area for a filed as well as if you mouse over the description itself. This hides the title for the field below. Unless there's a reason to read it while filling out the field, or to copy/paste the descriptions contents I think a better UX would be to only show when mousing over the title. (@ptitzler option on this?) (it can be seen in the video in bullet 3)

Yeah this is the current behavior so I figured it was the UX we wanted, but I could see that being a little annoying. If @ptitzler has an opinion let us know!

> * Show hide icon css not quite right

Fixed

> * Form spacing changes on initial mouseover, see video below:

I'm not able to reproduce this? Maybe we can do a call to debug

> * some escaping is going wrong here:

Fixed

> * description tooltips are misaligned for code snippet fields, also there is no description for tags (dont know if that's intentional)

Fixed misalignment of tooltips and added tooltip to code field. I don't think tags already had that so not sure if I should add that in? 

> * css for this field looks off

Fixed the spacing up a little, let me know if you have more input on the style

> * there's no longer button highlighting when mousing over the save button like before

Fixed

ptitzler commented:
It appears that the required annotation (*) isn't displayed for all required properties, as seen in this code snippet example. Note the discrepancy between the validation messages and the property annotations:

![image](https://user-images.githubusercontent.com/13068832/170332876-cc59bb3e-0125-44f7-b4d6-812041703a27.png)

ptitzler commented:
In various forms the label styling is inconsistent. For example (small, regular font vs bigger, bold font): 
![image](https://user-images.githubusercontent.com/13068832/170336620-961bf8de-8f89-4f61-accb-f00db616fe2e.png)

ptitzler commented:
It appears that for list-valued properties there are two new buttons displayed (move up, move down) that appear to give the user the ability to reorder list items.

![image](https://user-images.githubusercontent.com/13068832/170336911-b151d98e-240e-41db-8ba8-9cd801c6be4a.png)

Is this something that could be disabled? While there might be general use cases where order matters, I can't think of one for any of our properties. 

Part of the problem is that the list-specific action buttons are rendered in a way that feels too attention grabbing, considering that they don't serve a practical purpose.

![image](https://user-images.githubusercontent.com/13068832/170337995-073ed86a-831c-4457-a4ea-4a4249d5e4b4.png)

The old/existing rendering (shown here for a list-valued pipeline property) of list-specific blends in nicely and doesn't require additional screen real estate:

![image](https://user-images.githubusercontent.com/13068832/170338487-8ddca9d9-659a-4a66-a294-58dc6faa5d9a.png)

ptitzler commented:
Ageneral observation that the current styling appears to use almost always bold fonts (section headers, property labels, ...) as shown here for runtime configurations:

![image](https://user-images.githubusercontent.com/13068832/170339441-87973527-1883-49a4-bb8b-3d198538b4dd.png)

As a result everything seems to blend in, eliminating the benefits, for example, of section headers that were meant to help the user put labels into proper context. For comparison, here's the current rendering, which is a lot less attention grabbing:

![image](https://user-images.githubusercontent.com/13068832/170341644-1e68f71a-d06a-4804-b616-20eca99fc113.png)

ajbozarth reviewed (approved):
This LGTM, I gave @marthacryan a couple UI tweaks to follow up on over a Webex that I would consider non-blocking
